specify fixed-point arithmetic for rate calculations

avahowell opened the pull request:
this PR specifies fixed-point arithmetic for computing:

* base exchange rate
* commission rate
* validator reward rate
* validator exchange rate
* validator voting power adjustment

by adding an extra section to the Staking and Delegation part of the spec called "arithmetic".

Closes #271

avahowell commented:
the math here is in the same notation as the main part of the spec, but I also have a python implementation that I wrote in parallel to test things out, happy to include that in the spec as well if you think it would be helpful

hdevalence commented on spec/stake/arithmetic.md:
```diff
@@ -0,0 +1,59 @@
+
+The base exchange rate, $\psi(e)$, can be safely computed as follows:
+
+$$\psi(e) = \prod_{0 \leq i < e} \lfloor \frac{(1e8 + r_i)}{1e8} \rfloor $$
```
Should this be something like
```
$$\psi(e) = \left\lfloor \frac {\psi(e-1) (10^8 + r_e)} {10^8} \right\rfloor$$
```
instead? The existing formula doesn't seem quite right, because (1) if we apply the rounding to the rate term, we'll be rounding to 1 and then taking a product of 1s and (2) we know we have to compute the rate iteratively in order to have consistent rounding behavior.

avahowell replied:
good catch, I actually ended up implementing it this way in python but the change didn't make it back into the notation here

hdevalence commented on spec/stake/arithmetic.md:
```diff
@@ -0,0 +1,59 @@
+This allows outputs to fit in a u64, with all products fitting in the output
+and plenty of headroom for additions.
+
+All integer values should be interpreted as unsigned 64-bit integers, with the
+exception of the validator's commission rate, which is a `u16` specified in
+terms of basis points.
```
Just saying that they're `u64`s isn't enough to specify the interpretation, since we also need to specify that they have an implicit denominator of $10^8$.

I think it'd be useful to add a paragraph or something describing the interpretation of a `u64` integer value, with an example of what happens during multiplication: the value $x$ is represented as the integer $x 10^8$, so the product $xy$ is represented as the integer $xy 10^8$, which we compute from the representatives $x 10^8$ and $y 10^8$ as $\lfloor (x 10^8) (y 10^8) \cdot 10^{-8} \rfloor$. 

Maybe there's a better way to explain this than the above, but it'd be useful to have an explanation of why we're dividing by 10^8, it's because when we multiply two representatives, we get an extra factor of 10^8 that we have to divide out.

hdevalence commented on spec/stake/arithmetic.md:
```diff
@@ -0,0 +1,59 @@
+the ith rate at epoch $e$.  $s_i$ should be a fixed-precision `u16` integer
+type, specified in terms of [basis
+points](https://en.wikipedia.org/wiki/Basis_point). The resulting $c_{v,e}$ is
+defined in terms of basis points (i.e., four digits of precision). 
```
nit: this phrasing is slightly mixing between the units of measurement and the bits of precision; in our case, we're using integer values, so these are connected, but in principle the two concepts are doing different things (you could have only a few digits of precision, but for some arbitrary base quantity)

hdevalence commented on spec/stake/arithmetic.md:
```diff
@@ -0,0 +1,59 @@
+
+To compute the validator's base reward rate, we compute the following:
+
+$$r_{v,e} = \lfloor \frac{(1e8 - (c_{v,e}*1e4))r_e}{1e8} \rfloor$$
```
Here and elsewhere, could we do `10^8` rather than `1e8` in the math formulas?

avahowell commented:
updated the spec:

* fixed the formulas to be incremental
* added more information about the scheme, value representations, and a table of contents
